launcher/launcher.go

```diff
@@ -36,3 +63,805 @@ type NodeInfo struct {
+	defer launcher.Unlock()
+
+	launcher.resetDeviceAllocation()
+
+	launcher.currentDesiredInstances = instances
+	launcher.pendingNewServices = newServices
+	launcher.currentErrorStatus = launcher.performNodeBalancing(instances)
+
+	for _, node := range launcher.nodes {
+		node.waitStatus = true
+
+		launcher.nodeManger.RunInstances(
+			node.NodeID, node.currentRunRequest.services, node.currentRunRequest.layers, node.currentRunRequest.instances,
+			false)
+	}
+
+	return nil
+}
+
+// RestartInstances performs restart service instances.
+func (launcher *Launcher) RestartInstances() error {
+	launcher.Lock()
+	defer launcher.Unlock()
+
+	for _, node := range launcher.nodes {
+		launcher.initNodeUnitConfiguration(node, node.NodeType)
+	}
+
+	launcher.currentErrorStatus = launcher.performNodeBalancing(launcher.currentDesiredInstances)
+
+	for _, node := range launcher.nodes {
+		node.waitStatus = true
+
+		launcher.nodeManger.RunInstances(
+			node.NodeID, node.currentRunRequest.services, node.currentRunRequest.layers, node.currentRunRequest.instances,
+			true)
+	}
+
+	return nil
+}
+
+// GetInstanceStatsesChanhel gets channel with instances status.
+func (launcher *Launcher) GetInstanceStatsesChanhel() <-chan []cloudprotocol.InstanceStatus {
+	return launcher.runInstacnesStatusChannel
+}
+
+// GetNodesConfiguration gets nodes configuration.
+func (launcher *Launcher) GetNodesConfiguration() []cloudprotocol.NodeInfo {
+	nodes := make([]cloudprotocol.NodeInfo, len(launcher.nodes))
+
+	i := 0
+
+	for _, v := range launcher.nodes {
+		nodes[i] = v.NodeInfo.NodeInfo
+		i++
+	}
+
+	return nodes
+}
+
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+func (launcher *Launcher) fillUIDPool() {
+	uids, err := launcher.storage.GetAllUIDs()
+	if err != nil {
+		log.Errorf("Can't fill UID pool: %v", err)
+	}
+
+	for _, uid := range uids {
+		if err = launcher.uidPool.AddID(uid); err != nil {
+			log.Errorf("Can't add UID to pool: %v", err)
+		}
+	}
+}
+
+func (launcher *Launcher) processChannels(ctx context.Context) {
+	for {
+		select {
+		case instances := <-launcher.nodeManger.GetRunInstancesStatusChannel():
+			launcher.processRunInstanceStatus(instances)
+
+		case <-launcher.connectionTimer.C:
+			log.Errorf("Nodes connection timeout")
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (launcher *Launcher) processRunInstanceStatus(runStatus NodeRunInstanceStatus) {
+	launcher.Lock()
+	defer launcher.Unlock()
+
+	log.Debugf("Received runstatus from nodeID: %s", runStatus.NodeID)
+
+	currentStatus := launcher.getNode(runStatus.NodeID)
+	if currentStatus == nil {
+		if !slices.Contains(launcher.config.SMController.NodeIDs, runStatus.NodeID) {
+			log.Errorf("Received status for unknown nodeID  %s", runStatus.NodeID)
+			return
+		}
+
+		var err error
```

Comment:
why do we need to declare error variable here and initialize beloved

Reply:
because we reuse currentStatus  if i will use := in this case it will be new wariabel